File: LevelOrderTraversalOfBinaryTree/level_order_traversal_of_binary_tree.py
#!/usr/bin/env python3

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        from collections import deque
        
        queue = deque()
        if root: queue.appendleft(root)
        
        res = []
        
        while queue:
            level_vals = []

            for _ in range(len(queue)):
                node = queue.pop()
                level_vals.append(node.val)
                if node.left: queue.appendleft(node.left)
                if node.right: queue.appendleft(node.right)
            res.append(level_vals)
        return res

File: LevelOrderTraversalOfBinaryTree/level_order_traversal_of_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        deque<TreeNode*> queue;

        if ( root ) queue.push_back( root );

        while ( !queue.empty() ){
            vector<int> level_vals {};

            for ( int i = 0, len = queue.size(); i != len; i++ ){
                TreeNode* node {queue.front()};
                queue.pop_front();
                level_vals.push_back(node->val);

                if (node->left) queue.push_back(node->left);
                if (node->right) queue.push_back(node->right);
            }
            res.push_back(level_vals);
        }
        return res;
    }
};

